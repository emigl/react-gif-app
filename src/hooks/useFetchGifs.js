import { useEffect, useState } from 'react'
import getGifs from '../helpers/getGifs'


export const useFetchGifs = (category) => {

    const [state, setState] = useState({
        data: [],
        loading: true
    })

    useEffect(() => {

        getGifs(category)
            .then(images => {
                setState({
                    data: images,
                    loading:false
                })
            })

    },[category])

    return state;
}

