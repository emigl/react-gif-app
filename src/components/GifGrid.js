import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
import PropTypes from 'prop-types'

// import getGifs from '../helpers/getGifs';
import GifGridItem from './GifGridItem';

const GifGrid = ({category}) => {

    // Al desestructurar, se puede renombrar de la siguiente forma, primero con el nombre que recoge el dato a desestructurar y después dos puntos y el alias, es decir data:alias
    const {data:images, loading} = useFetchGifs(category);

    // useEffect ejecuta el código dentro de la funcion cuando el componente es renderizado por primera vez, dentro de la funcion recibe el código que se quiere disparar al renderizarlo, después como segundo parametro de la funcion, se le pasa un array de dependencias, es decir, cuando cambia alguna dependencia, disparar el codigo de nuevo
    // useEffect(() => {
    //     getGifs(category)
    //     .then(images => setImages(images))
    // },[category])
    
   
  
    return (
        <div className="gifs-container">
            {loading && <h2>Loading</h2>}
            <h3 className="category-title animate__animated animate__fadeIn">{ category } </h3>

            
                { images.map( (image) => {
                    return (
                       <GifGridItem
                        key={image.id} 
                        // Con el operador spread utilizado de esta forma es posible mandar las props a un componente de forma independiente, para que se pueda desestructurar más facilmente
                        {...image} 
                        />
                    )
                })}
            
        </div>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}

export default GifGrid
