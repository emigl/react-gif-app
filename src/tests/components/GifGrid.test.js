
import {shallow} from 'enzyme'
import GifGrid from '../../components/GifGrid'
import { useFetchGifs } from '../../hooks/useFetchGifs'
// mock() fingirá que tiene la data que trae useFetchGifs sin llamarla directamente
jest.mock('../../hooks/useFetchGifs')

describe('Test de GifGrid component', () => {
    
    const category = 'Evangelion'

    test('debera renderizar correctamente con el snapshot', () => {
        // Se usa de esta forma el mockReturnValue y se pasa por parámetro el mismo objeto que espera los componentes que llaman a el metodo useFetchGifs
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        })
        const wrapper = shallow(<GifGrid category={category}/>)

        expect(wrapper).toMatchSnapshot()
        
    })

    test('debe de mostrar items cuando se cargan imagenes useFetchGifs', () => {
        
        const gifs = [{
            id: '231e',
            url: 'https://localhost/cualquiercosa.png',
            title: 'Esto es un titulo'
        },
        {
            id: 'njhubnu',
            url: 'https://localhost/c2.png',
            title: 'Esto es 2'
        },
        {
            id: '1231288',
            url: 'https://localhost/c3.png',
            title: 'Esto es 3'
        }]

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        })
        const wrapper = shallow(<GifGrid category={category}/>)

        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('h2').exists()).toBe(false)
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length)
     
    })
    
    
})
