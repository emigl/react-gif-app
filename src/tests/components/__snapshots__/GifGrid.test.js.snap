// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test de GifGrid component debe de mostrar items cuando se cargan imagenes useFetchGifs 1`] = `
<div
  className="gifs-container"
>
  <h3
    className="category-title animate__animated animate__fadeIn"
  >
    Evangelion
     
  </h3>
  <GifGridItem
    id="231e"
    key="231e"
    title="Esto es un titulo"
    url="https://localhost/cualquiercosa.png"
  />
  <GifGridItem
    id="njhubnu"
    key="njhubnu"
    title="Esto es 2"
    url="https://localhost/c2.png"
  />
  <GifGridItem
    id="1231288"
    key="1231288"
    title="Esto es 3"
    url="https://localhost/c3.png"
  />
</div>
`;

exports[`Test de GifGrid component debera renderizar correctamente con el snapshot 1`] = `
<div
  className="gifs-container"
>
  <h2>
    Loading
  </h2>
  <h3
    className="category-title animate__animated animate__fadeIn"
  >
    Evangelion
     
  </h3>
</div>
`;
